blog/_posts/2019-11-14-artifacts.md

```diff
@@ -0,0 +1,232 @@
+
+# Updating your packages
+
+To update your packages, first build with the latest version of `BinaryBuilder`, then add the autogenerated JLL packages as dependencies in your projects and packages, use the new API for your `ccall()`'s and `run()`'ing your binaries, and give a small smile as you delete your `deps/build.jl` file.  Death to all global mutable state.
+
+## Beyond JLL packages
+
+For an example of how JLL packages can be used for more than just JLL packages, we look to an [example in `Gtk.jl`](https://github.com/staticfloat/Gtk.jl/blob/e500dcadaa8bf3571d27b0ac7cd92ef00febcb1f/src/Gtk.jl#L82-L111):
+
+```julia
+mutable_artifacts_toml = joinpath(dirname(@__DIR__), "MutableArtifacts.toml")
+loaders_cache_name = "gdk-pixbuf-loaders-cache"
+loaders_cache_hash = artifact_hash(loaders_cache_name, mutable_artifacts_toml)
+if loaders_cache_hash === nothing
+    # Run gdk-pixbuf-query-loaders, capture output, 
+    loader_cache_contents = gdk_pixbuf_query_loaders() do gpql
+        withenv("GDK_PIXBUF_MODULEDIR" => gdk_pixbuf_loaders_dir) do
+            return String(read(`$gpql`))
+        end
+    end
+
+    # Write cache out to file in new artifact
+    loaders_cache_hash = create_artifact() do art_dir
+        open(joinpath(art_dir, "loaders.cache"), "w") do io
+            write(io, loader_cache_contents)
+        end
+    end
+    bind_artifact!(mutable_artifacts_toml,
+        loaders_cache_name,
+        loaders_cache_hash;
+        force=true
+    )
+end
+
+# Point gdk to our cached loaders
+ENV["GDK_PIXBUF_MODULE_FILE"] = joinpath(artifact_path(loaders_cache_hash), "loaders.cache")
+ENV["GDK_PIXBUF_MODULEDIR"] = gdk_pixbuf_loaders_dir
+```
+
+Within the `__init__()` method of `Gtk`, this is looking to see if a local cache of `gdk-pixbuf` loaders has been generated.  This cache is machine-specific, and must be generated upon first run of the module.  Doing so requires invoking a binary from the `gdk_pixbuf_jll` package, which is done using that JLL package's `gdk_pixbuf_query_loaders` export.  By writing the result out to a new artifact and binding that artifact to a `MutableArtifacts.toml` file (an arbitrarily-named `.gitignore`'d file), we are able to dynamically cache binary objects that are kept separately from the rest of our package data.  Gtk is thereafter informed of the cache location through environment variables.
```

Comment:
Storing configuration next to the source file should be considered bad practice imo.

Reply:
I totally agree with you, but since we don't have `Pkg.cache()` yet (https://github.com/JuliaLang/Pkg.jl/issues/796#issuecomment-523154714) this is the only location that a package can stash files without potentially conflicting with other versions of itself at the moment.

Reply:
Maybe mention that there is a plan to improve this in the future?